src/utils/mutex.ts

```diff
@@ -0,0 +1,52 @@
+import {logWarn} from './log';
+
+/**
+  * This is the Mutex(Not to confuse with RW-Mutex).
+  * This is to ensure that certain access are only done by 1 callback(a current fallpit of PromiseBarrier).
+  * This ensures that their is a global mutex which can be called by a finite amount of callbacks.
+  * We should ensure that the mutex can only be locked by the first callback which has requrested it
+  * and only give the next callback the mutex lock once it's been released.
+ */
+export class Mutex {
+    private awaitingResolves = [] as Array<() => void>;
+    private locked = false;
+
+    public isLocked() {
+        return this.locked;
+    }
+
+    // Request a lock from the mutex.
+    public async lock() {
+        // Easiest path, the mutex is not locked.
+        // We lock the mutex and
```

Comment:
> We lock the mutex and

Such a cliffhanger. Let's finish the sentence. :)

Reply:
Ah sigh, not sure what VSCode did with my comments today. Maybe time to search for alternatives.